### ISSUE TITLE
[Distributed Index] Roles not assigning channels to users correctly

### ISSUE THREAD

**Issue by sethrosetter** [couchbase/sync_gateway#1307]
Test demonstrating the failure: https://github.com/couchbaselabs/sync-gateway-testcluster/blob/sync-scenarios-1/functional_tests/test_roles.py

TLDR
A role ("radio_stations") is created with (["KMOW", "HWOD", "KDWB"]) channels
A user is assigned to that role
Docs are pushed with channels set to "KMOW" or "HWOD" or "KDWB"
The user issues a _changes request

Expected: The user would get the docs for all the channels in the role

Actual (On feature/distributed_index): The user gets 0 docs
Actual (On master): The user gets all of the docs for "radio_stations" role

The sync_function is default.
### On master

GET 4985/db/_role/radio_stations

```
{
"name": "radio_stations",
"admin_channels": [
"HWOD",
"KDWB",
"KMOW"
],
"all_channels": [
"!",
"HWOD",
"KDWB",
"KMOW"
]
}
```

GET 4985/db/_user/dj_0

```
{
"name": "dj_0",
"all_channels": [
"!",
"HWOD",
"KDWB",
"KMOW"
],
"admin_roles": [
"radio_stations"
],
"roles": [
"radio_stations"
]
}
```
### On "Couchbase Sync Gateway/feature/distributed_index(383c756)"

GET 4985/db/_role/radio_stations

```
{
"name": "radio_stations",
"admin_channels": [
"HWOD",
"KDWB",
"KMOW"
],
"all_channels": [
"!",
"HWOD",
"KDWB",
"KMOW"
]
}
```

GET 4985/db/_user/dj_0

```
{
"name": "dj_0",
"all_channels": [
"!"
],
"admin_roles": [
"radio_stations"
]
}
```

It looks like roles and all_channels are not getting assigned to the user correctly in the distributed_index branch

**Comment by adamcfraser** [comment]
I'm not able to reproduce against the distributed_index branch - when I create a role, then a user that's assigned to that role using admin_roles, a subsequent GET on the user shows the full list of channels from the role.

I tested doing the following (httpie):

```
http PUT localhost:4985/default/_role/testRole name=testRole admin_channels:='["ABC"]'

http PUT localhost:4985/default/_user/roleUser name=roleUser password=1234 admin_roles:='["testRole"]'

http GET localhost:4985/default/_user/roleUser

Response:
{"name":"roleUser","all_channels":["!","ABC"],"admin_roles":["testRole"],"roles":["testRole"]}


Are you able to reproduce via curl/httpie for the failing setup of user and role, or at least observe the calls being made by the test?   I'm not familiar enough with the test suite to know what API calls register_bulk_users is making.
```

**Comment by adamcfraser** [comment]
When there's no global integer sequence, explicit role assignment is being skipped by seq>0 handling in TimedSet.AddAtSequence.

Fix should be in users.go, in UpdatePrincipal - if dbc.writeSequences() == false, use nextSeq!=0, to force role and channel assignments.

Should be done in conjunction with #993, to ensure appropriate value is used.

**Comment by adamcfraser** [comment]
Should be addressed by #1393.  Assigning to QE for verification.

**Comment by sethrosetter** [comment]
@adamcfraser This is not fixed post merge down. http://qa.sc.couchbase.com/view/mobile/job/centos-syncgateway-functional-tests/43/testReport/junit/functional_tests/test_roles/test_roles_sanity/

**Comment by sethrosetter** [comment]
@adamcfraser This test failed against build 1.2.0-17. Was this pre or post access work? 

It is no longer failing as of 1.2.0-20.

 I'm going to put this back in your hands. Please close if this was resolved between those two builds.

**Comment by adamcfraser** [comment]
Commit with the fix is:
https://github.com/couchbase/sync_gateway/commit/ceef61cad53554018ef9b2c1479a124721b11b3d

That looks like 1.2.0-20 to me.

**Comment by adamcfraser** [comment]
I don't think this was an intermittent failure, though - if it's passing on the latest master build, that's probably all we care about.

**Comment by sethrosetter** [comment]
@adamcfraser cool. Seems resolved then. I haven't seen it fail since 1.2.0-20
